test/e2e/network/dual_stack.go

```diff
@@ -205,6 +206,161 @@ var _ = SIGDescribe("[Feature:IPv6DualStackAlphaFeature] [LinuxOnly]", func() {
+		if framework.TestContext.IPFamily == "ipv4" && !isIPv4(result.Spec.ClusterIP) {
+			framework.Failf("got unexpected service ip %s, primary service ip family %s", result.Spec.ClusterIP, framework.TestContext.IPFamily)
+		}
+		// ensure endpoints belong to same ipfamily as service
+		endpoints, err := cs.CoreV1().Endpoints(service.Namespace).Get(service.Name, metav1.GetOptions{})
+		if err != nil {
+			framework.Failf("Get endpoints for service %s/%s failed (%s)", service.Namespace, service.Name, err)
+		}
+		if len(endpoints.Subsets) == 0 {
+			framework.Failf("Endpoint has no subsets, cannot determine service ip family matches endpoints ip family.")
+		}
+		for _, ss := range endpoints.Subsets {
+			for _, e := range ss.Addresses {
+				if framework.TestContext.IPFamily == "ipv4" && !isIPv4(e.IP) {
+					framework.Failf("service ip %s doesn't belong to ip family %s", e.IP, framework.TestContext.IPFamily)
+				}
+			}
+		}
+	})
+
+	ginkgo.It("should create service with ipv4 cluster ip [Feature:IPv6DualStackAlphaFeature:Phase2]", func() {
+		serviceName := "ipv4clusterip"
+		ns := f.Namespace.Name
+		ipv4 := v1.IPv4Protocol
+
+		jig := e2eservice.NewTestJig(cs, serviceName)
+
+		t := e2eservice.NewServerTest(cs, ns, serviceName)
+		defer func() {
+			defer ginkgo.GinkgoRecover()
+			errs := t.Cleanup()
+			if len(errs) != 0 {
+				framework.Failf("errors in cleanup: %v", errs)
+			}
+		}()
+
+		ginkgo.By("creating service " + serviceName + " with Service.Spec.IPFamily IPv4" + ns)
+		service := createService(t.ServiceName, t.Namespace, t.Labels, &ipv4)
+
+		jig.Labels = t.Labels
+		jig.CreateServicePods(cs, ns, 2)
+		result, err := t.CreateService(service)
+		framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
+
+		if len(result.Spec.Ports) != 2 {
+			framework.Failf("got unexpected len(Spec.Ports) for new service: %v", result)
+		}
+		// check the spec has been set to IPv4
+		if result.Spec.IPFamily == nil || *result.Spec.IPFamily != v1.IPv4Protocol {
+			framework.Failf("got unexpected service ip family for new service: %s", *result.Spec.IPFamily)
+		}
+		// check to ensure cluster ip is ipv4
+		if !isIPv4(result.Spec.ClusterIP) {
+			framework.Failf("got unexpected service ip %s, should be ipv4 ip", result.Spec.ClusterIP)
+		}
+		// ensure endpoints belong to same ipfamily as service
+		endpoints, err := cs.CoreV1().Endpoints(service.Namespace).Get(service.Name, metav1.GetOptions{})
+		if err != nil {
+			framework.Failf("Get endpoints for service %s/%s failed (%s)", service.Namespace, service.Name, err)
+		}
+		if len(endpoints.Subsets) == 0 {
+			framework.Failf("Endpoint has no subsets, cannot determine service ip family matches endpoints ip family.")
+		}
+		for _, ss := range endpoints.Subsets {
+			for _, e := range ss.Addresses {
+				if !isIPv4(e.IP) {
+					framework.Failf("service endpoint %s doesn't belong to ipv4 ip family", e.IP)
```

Comment:
What would cause this to fail in the code? Do we explicitly allow intermixing when adding to an endpoint set? Is this something that different cloud-providers might leak in?

Reply:
When adding to endpoint set, there is no intermixing. Based on the ip family, the endpoints are added. This is an additional sanity check to ensure the behavior is as intended.

Reply:
Great that should help me close out a few issues related to this problem currently.